### ISSUE TITLE
Feature request: sanitizing user-provided params

### ISSUE THREAD

**Issue by SynedraAcus** [nextflow-io/nextflow#4878]
## Sanitizing user-provided params

In Nextflow as it currently stands, any param that is used in a process gets copied to `.command.sh` without validation. Although filenames are properly processed (*eg* spaces are correctly quoted by backslashes), string parameters are not; instead, they are inserted into process script verbatim. This can both break processes if parameters contain characters meaningful for Bash (even non-quoted spaces break processes written without proper precautions) and, more importantly, allow executing arbitrary code at the machine where Nextflow runs.

## Issue demonstration

This POC pipeline is enough to demonstrate:

`./main.nf`:
  
```
  params.param1 = 'lorem ipsum'
  params.param2 = 'dolor sit amet'
  
  process PASS_UNSCREENED{
      output: stdout
      """
      echo "Not screened"
      ${moduleDir}/check_params.py -f ${params.param1} -o ${params.param2}
      """
  }
  
  process PASS_SCREENED{
      output: stdout
      """
      echo "Screened"
      ${moduleDir}/check_params.py -f '${params.param1}' -o '${params.param2}'
      """
  }
  
  workflow{
      PASS_UNSCREENED | view()
  
      PASS_SCREENED | view()
  }
```

`check_params.py`:
  
```
  #! /usr/bin/env python3
  from argparse import ArgumentParser
  parser = ArgumentParser()
  parser.add_argument('-f', nargs='*', type=str)
  parser.add_argument('-o', type=str)
  args = parser.parse_args()
  print('Multivalue param:')
  for i in args.f:
    print(i)
  print('Single value param:')
  print(args.o)

```
Unsurprisingly, wrapping `${params.param_name}` calls in single quotes prevents Bash expansion and the most basic form of code injection:

<img width="947" alt="inject" src="https://github.com/nextflow-io/nextflow/assets/7446382/2ae22bc4-8d53-4ca3-8911-43f82916dfed">

However, it is not enough to protect from intentionally created malicious input. Something as simple as `--param1 "asd ' && echo $Path '"` will successfully execute the payload. Since parameters are typically provided by end users (or upstream pipelines) and are out of Nextflow administrators' control, this presents a security risk.

## Proposed solution

There could be an option for checking all user-provided parameters for whether they are safe to use with Bash. Something like `nextflow run main.nf -params-file user_supplied.json -sanitize -replace-with '_'` that would allow either exiting immediately if forbidden characters are encountered in params, or replacing all such characters with, for example, an underscore. Since it is possible that existing pipelines rely on using the now-forbidden characters in some of the parameters, this has to be optional.

**Comment by bentsherman** [comment]
It's not so simple because process scripts can be written in any scripting language, not just bash, so the sanitization will depend on the interpreter.

It could be interesting to implement something similar to [this proposal](https://openjdk.org/jeps/430) to add string templates to Java with a new concept of "template processors" which define the sanitization for a particular output, such as plain-text, CSV, HTML, SQL, etc. The sanitization should be built into the pipeline code rather than as a command line option.

I think the main attack vector is when you run a pipeline and include some remote config which could execute arbitrary code, or define params which execute arbitrary code in the task script.

**Comment by SynedraAcus** [comment]
I agree that it is possible and useful to define a Groovy process that does sanitizing, like [this SO answer](https://stackoverflow.com/questions/36426677/removing-special-characters-from-a-string-in-a-groovy-script). Or pre-process values on the caller side, if there is an additional layer of code between users and Nextflow. But for our team (just to give an example), the question of whether parameters require some kind of processing wasn't raised until well into development; I suspect it just never occured to a lot of people. Having a reasonable default or an easily available (and visible!) filtering option should patch this particular hole.

As for the language specifics, it is complicated but possible. A user-configurable list of forbidden characters? A number of distinct sanitization functions for Bash, Python, Perl, *etc.*, perhaps as a process parameter?
